runtime/interpreter/interpreter.go

```diff
@@ -978,6 +980,47 @@ func (interpreter *Interpreter) declareAttachmentValue(
 	return interpreter.declareCompositeValue(declaration, lexicalScope)
 }
 
+// evaluates all the implicit default arguments to the default destroy event
+//
+// the handling of default arguments makes a number of assumptions to simplify the implementation;
+// namely that a) all default arguments are lazily evaluated at the site of the invocation,
+// b) that either all the parameters or none of the parameters of a function have default arguments,
+// and c) functions cannot currently be explicitly invoked if they have default arguments
+//
+// if we plan to generalize this further, we will need to relax those assumptions
+func (interpreter *Interpreter) evaluateDefaultDestroyEvent(
+	containerComposite *CompositeValue,
+	compositeDecl *ast.CompositeDeclaration,
+	compositeType *sema.CompositeType,
+	locationRange LocationRange,
+) (arguments []Value) {
+	parameters := compositeDecl.DeclarationMembers().Initializers()[0].FunctionDeclaration.ParameterList.Parameters
+
+	interpreter.activations.PushNewWithParent(interpreter.activations.CurrentOrNew())
+	defer interpreter.activations.Pop()
+
+	var self MemberAccessibleValue = containerComposite
+	if containerComposite.Kind == common.CompositeKindAttachment {
+		var base *EphemeralReferenceValue
+		base, self = attachmentBaseAndSelfValues(interpreter, containerComposite)
+		interpreter.declareVariable(sema.BaseIdentifier, base)
+	}
+	interpreter.declareVariable(sema.SelfIdentifier, self)
+
+	for _, parameter := range parameters {
+		// lazily evaluate the default argument expressions
```

Comment:
In what sense is the evaluation "lazy" here? It looks like all arguments are eagerly evaluated

Reply:
They are "lazy" with respect to the event's declaration: if i declare a default event `ResourceDestroyed(foo: Int = self.x)`, `self.x` is evaluated in the context that exists when the event is destroyed, not the context when it is declared. This function is only called after the destroy triggers the event emission, so with respect to this function it's "eager".

Reply:
Ah, I see! Could you please add a (variant of) this explanation as a comment?